### ISSUE TITLE
Account for unfriendly characters in filenames when using field_for_media_title (and document this settings)

### ISSUE THREAD

**Issue by mjordan** [mjordan/islandora_workbench#613]
There is an inconsistency between https://github.com/mjordan/islandora_workbench/blob/main/workbench_utils.py#L3125 and https://github.com/mjordan/islandora_workbench/blob/main/workbench_utils.py#L5873-L5875 that needs to be fixed. Also, we probably should replace all non-filename-friendly characters in the target string with `_`, not just `:`.

Finally, this setting is not documented.

Tagging @dmer and @ajstanley in case they have any input.

**Comment by mjordan** [comment]
@dmer can you test this at your earliest convenience?

**Comment by mjordan** [comment]
No problem. But I'm confused - I almost certainly rushed into fixing this without thinking very much. This config setting is used in two places: 1) to assign the media's title (Drupal calls it the media's "name") to the value of the defined CSV field, and 2) to assign the media file's name using that field. I don't think we need to remove non-alphanumeric characters from the media title, do we? Just the filename? Also, maybe we want to have two separate config settings, one for each?

**Comment by dmer** [comment]
Ah, I wasn't paying attention to that difference - in this case the repository owner wants both the media title and the file name to be the same - derived from their field_identifier column (this is their overall file naming convention). I'm not sure that I've encountered any situations yet where folks wanted that media title to be different from the filename. I guess I'm of two minds - on the one hand I agree that there's no reason to remove the "illegal" chars from the title - and on the other hand it seems like it might lead to confusion if the title and filename are not identical. I'll ask around here at BD if anyone has a strong opinion and invite them to post such here.

**Comment by mjordan** [comment]
See related comment in #615.

**Comment by dmer** [comment]
@mjordan I don't have a strong preference - so no objections to adding that Media_title field.

**Comment by dara2** [comment]
Hi @mjordan - I also had an issue perhaps similar to what Liza reported over in #615. I am using the `field_for_media_title` config and I am pulling remote files from an i7 site. For files that had ".tif" extensions, they look like they were ingested, but if I click on the actual file it gives a 404, and no derivatives were made. I'm wondering if this is a totally different problem, or it relates to the combo of the remote file extensions code and the `field_for_media_title` configuration? (I don't see any "ERROR" messages in the workbench.log during the ingest.)
Here's an example of one of the files that didn't ingest: https://diglib.amphilsoc.org/islandora/object/text%3A258914/datastream/OBJ/view. Its label in i7 has an extension of ".tif" but when I download it myself, it gets a ".tiff" extension. But in the i2 site, it has a ".tif" extension and doesn't actually exist (see attached screenshot).

![Screenshot 2023-07-03 at 4 54 10 PM](https://github.com/mjordan/islandora_workbench/assets/72275887/577e6ef3-3b75-41a2-9f6a-ec067014741c)

**Comment by dara2** [comment]
^ I can move this over to #615 if that makes more sense, but put it here since Liza's issue got moved over here and 615 closed. Or I can open a new issue! :)

**Comment by mjordan** [comment]
Hi @dara2, I won't know until I investigate. I'll try to do so today.

**Comment by dara2** [comment]
Hi Mark - I wondered if you might have time to look at this this week? We've been having a lot of trouble pulling remote files (which didn't used to be problematic) which is delaying being able to migrate a client into i2. Let me know if there's anything we can check on our end that might be helpful too - thank you!

**Comment by mjordan** [comment]
HI @dara2, yes, I can continue with this work this week. I will be providing a description of what I have in mind here this evening (no major changes, just a couple new config settings).

**Comment by mjordan** [comment]
I would like to report out on some investigation around the naming of downloaded remote media files (and the related media names/titles) and propose some additional config settings in this area.

Finding 1: `field_for_media_title` and `use_nid_in_media_title` are currently undocumented (as @dmer discovered). I  propose keeping these two settings and documenting them as:

* `field_for_media_title`: if defined in config, media titles will be based on the defined field, with : replaced by _ [this is the current behavior - @ajstanley I think you had the original use case for this; I have forgotten why we decided to replace any : with _. Could we just take the field value verbatim?]
* `use_nid_in_media_title`: if set to true, the media title will be `{node_id}-Original File`. [I am open to allowing users to define their own wording for this media title using a simple template system; if interested, please say so here.]

Finding 2: there are two currently documented settings relevant to media and remote file naming. These can stay as is:

* `use_node_title_for_media_title`: "If set to true (default), name media the same as the parent node's title value. If false, the filename of the media will be used as the media name. Applies to both create and add_media tasks."
* `use_node_title_for_media` :"For remote files, if set to true [false is default], name the downloaded copy using the parent node's title value. If false, downloaded file's name will depend on the application/website that is serving up the file. Applies to create tasks only."

Proposal 1: rename `use_node_title_for_media` to `use_node_title_for_media_filename`, in the interests of clarity.

Proposal 2: add the following config settings: 

- `use_node_title_for_media_filename`: if set to true, the downloaed filename will be the node's title with all non-alphanumeric characters changed to _.
- `use_nid_in_media_filename`: if set to true, the downloaded filename will be {node_id}-Original File{remote_extension_with_dot}
- `field_for_media_filename`: if defined in config, the downloaded filename will be the taken from the fieldname provided in this setting, with all non-alphanumeric characters changed to _.

**Comment by dara2** [comment]
Hi Mark - Thank you for looking into all of this! I am trying to parse it all out, and one question I have is for this:
> use_node_title_for_media :"For remote files, if set to true [false is default], name the downloaded copy using the parent node's title value. If false, downloaded file's name will depend on the application/website that is serving up the file. Applies to create tasks only."

Does this only apply to the media name, or does it apply to the filename? I wasn't sure because of the phrase "If false, the downloaded file's name..." Maybe it could say "...the downloaded file's media name"?

**Comment by dara2** [comment]
I also wonder if it might be helpful to put "remote" in the config name for config options that just apply to remote files?

**Comment by mjordan** [comment]
@dara2 does the existing `use_node_title_for_media_title` do what you are asking about?

**Comment by dara2** [comment]
I was just talking about the naming conventions. I've had a hard time keeping them straight and understanding what each one does. So I wondered:

- Can the config names refer consistently to "media title" when it's talking about the title on the Media tab, and to "filename" when it's talking about the name of the file itself?
- Can configs that are used just for remote files maybe have "remote" in the name somewhere?

**Comment by mjordan** [comment]
Yes, those are good suggestions. What about these:

Currently exist but undocumented

- `field_for_media_title`: this currently applies only to media entities (not filenames) created both from local and remote files so we might want to leave it as is
-  `use_nid_in_media_title`: currently applies only to media created from remote files but this raises the question of whether anyone is interested in having it apply to media created from either local and remote files 

Currently exist and documented:

- `use_node_title_for_media_title`: this currently applies only to media entities (not filenames) 
- `use_node_title_for_media` (possibly renamed to `use_node_title_for_media_filename`):  currently applies only to media created from remote files but this raises the question of whether anyone is interested in having it apply to media created from either local and remote files

Propose to add:

- `use_node_title_for_media_filename`: rename to `use_node_title_for_remote_filename`?
- `use_nid_in_media_filename`: rename to `use_nid_in_remote_filename`?
- `field_for_media_filename`: rename to `field_for_remote_filename`?

**Comment by dara2** [comment]
I think that looks great!

**Comment by mjordan** [comment]
Super, thanks @dara2.

@ajstanley @dmer OK to go ahead with this? Note that the existing `use_node_title_for_media` will be renamed to `use_node_title_for_media_filename`, as above.

**Comment by dara2** [comment]
Mark - did you happen to figure out what might have been happening re: https://github.com/mjordan/islandora_workbench/issues/613#issuecomment-1619142493? It seemed like it was related to remote file extensions in combination with setting the media title. I wasn't sure if it needed its own issue?

**Comment by dmer** [comment]
Looks good to me @mjordan

**Comment by mjordan** [comment]
@dara2 I hope to include a fix for that (or at least an attempted fix) in this bunch of work.

**Comment by dara2** [comment]
Woohoo! That's awesome 😃

**Comment by mjordan** [comment]
Follow up question - in addition to the colon (`:`), what other filename-"unfriendly" characters should we be replacing with `_`? I'd suggest spaces and forward and backward slashes (`/` and `\`). Any others?

**Comment by mjordan** [comment]
Yes, good idea, we shouldn't allow any type of quotation mark. Thanks!

**Comment by dara2** [comment]
Hi Mark! I wondered if you might have time to get back to this this week?

**Comment by mjordan** [comment]
I've pushed a fresh `issue_613` branch that I'd like some real world testing of. Here are the new config settings:

Media titles:

- `field_for_media_title`:  Identifies a CSV field name (machine name, not human readable) that will be used as the media title in `create` tasks. Truncates the value of the field to 255 characters. For example, `field_for_media_title: id`.
- `use_nid_in_media_title`: Defaults to false. Set to true to use the parent node's node ID as the media title.
- `use_node_title_for_media_title`: Defaults to false. Set to true to use the parent node's title as the media title. Truncates the value of the field to 255 characters.

Remote filenames:

- `use_node_title_for_remote_filename`: (formerly `use_node_title_for_media_filename`) Defaults to false. Set to true to use a version of the parent node's title as the filename for a remote (http[s]) file. Truncates the value of the field to 255 characters. Replaces all non-alphanumeric characters with an underscore (`_`).
-  `use_nid_in_remote_filename`: (formerly `use_nid_in_media_filename`)  Defaults to false. Set to true to use the parent node's node ID as the filename for a remote (http[s]) file.
- `field_for_remote_filename`: (formerly `field_for_media_filename`) Identifies a CSV field name (machine name, not human readable) that will be used as the filename for a remote (http[s]) file. Truncates the value of the field to 255 characters. For example, `field_for_remote_filename: id`.

Thanks!

**Comment by dara2** [comment]
I would be happy to test this out this week, thank you!

**Comment by dara2** [comment]
(Working on testing this, but the i7 site I'm pulling data from has been down...stay tuned!)

**Comment by dara2** [comment]
It seems this might be related to https://github.com/mjordan/islandora_workbench/issues/593? But the MIME type for the problematic images, when I look at /manage/datastreams in i7, is image/tiff, so that shouldn't be a problem. I'm not sure, though, why the file's label in i7 is, for example, `Mss_B_F826-063.tif` - where would that ".tif" have come from?

**Comment by mjordan** [comment]
I'm not clear on what you mean by "label". Is it appearing in the datastreams list for the I7 object (example from our site below)?

![image](https://github.com/mjordan/islandora_workbench/assets/403918/e0d206b6-04e2-4679-999f-26214a99b2ee)

**Comment by mjordan** [comment]
Maybe the label is the original name of the file when it was uploaded to Islandora 7? When I `curl` https://diglib.amphilsoc.org/islandora/object/text:302158/datastream/OBJ/view the server doesn't return a filename of any kind. It does return a MIME type of `image/tiff`, which Workbench uses [to assign an extension](https://github.com/mjordan/islandora_workbench/blob/issue_613/workbench_utils.py#L5735) of `.tif` to the temporary copy it downloads.

I'll need to dig a bit deeper to understand why Workbench claims https://diglib.amphilsoc.org/islandora/object/text:302158/datastream/OBJ/view is not accessible.

**Comment by dara2** [comment]
(If you'd like me to create 2 separate issues for each of these, let me know. I feel like they diverge from what this issue's title describes but I don't want to confuse anything by moving them into different threads if it's easier to keep them here!)

**Comment by mjordan** [comment]
Thanks for the error lost, that's useful. All entries are showing that the server timed out (HTTP error code 522). So those files are accessible otherwise, it's just that requests for them are timing out. It's likely that every time you run Workbench `--check` against that server, the list of requests that returns the 522 response is not identical due to server load, etc. Making Workbench handle request timeouts is probably a separate issue but let's leave it aside for the moment.

I am able to use `curl` to retrieve https://diglib.amphilsoc.org/islandora/object/text:302175/datastream/OBJ/view. The MIME type provided by the server is `content-type: image/tiff`, and the file that `curl` downloads is indeed a TIFF file,  so I assume that the HTML that Workbench got when it tried to fetch the file was an error message page. If that was the case, the file that Workbench got in response to requesting that file was an error page that said "We're sorry, there was a problem, please try again later" or something similar. I just tried ingesting that file now using Workbench and it worked as expected:

![image](https://github.com/mjordan/islandora_workbench/assets/403918/69e85882-8557-47d4-a24a-bb8d663839ec)

However, I will add an entry to https://mjordan.github.io/islandora_workbench_docs/troubleshooting/ about HTML files being retrieved unexpectedly. I've seen this before - the server returns a human-friendly "error message" instead of the expected HTTP response code. It's one of those situations where humans and scripts have different expectations.

Let's focus on the .tif vs. .tiff. I think we can safely disregard the I7 Label value, since (at least in my experience) that label is intended to describe the label of the datastream. To my understanding, it's the human-readable version of the Datastream ID. I don't know why the site you're working with uses filenames there instead of the standard Islandora labels like the ones shown in my screenshot, but some plausible explanations are that someone used Datastream CRUD or a custom script to create the datastream and provided a custom value as the datastream Label. Without knowing more about the history of the site I don't think that the label you're seeing ends in .tiff and not .tif is that important. 

But, when you say 

> The extension I'm seeing on image/tiff files after ingest is .tiff, not .tif - noting this in case it's important.

Are you saying that Workbench is assigning .tiff as the extension to TIFFs that it retrieves from remote servers and ingests into Islandora 2? Workbench [should be assigning](https://github.com/mjordan/islandora_workbench/blob/issue_613/workbench_utils.py#L5735) .tif as the extension to TIFFs, not .tiff.

**Comment by mjordan** [comment]
Ha, https://mjordan.github.io/islandora_workbench_docs/troubleshooting/ already has an entry about unexpected reports of HTML files. The readability of that page certainly could use some improving. I'll add the specific error message to make it clearer.

**Comment by dara2** [comment]
Thank you SO MUCH for looking into this and for this information!

> Are you saying that Workbench is assigning .tiff as the extension to TIFFs that it retrieves from remote servers and ingests into Islandora 2? Workbench [should be assigning](https://github.com/mjordan/islandora_workbench/blob/issue_613/workbench_utils.py#L5735) .tif as the extension to TIFFs, not .tiff.

Yes, it's assigning .tiff as the file extension. Example:
![Screenshot 2023-07-21 at 10 54 23 AM](https://github.com/mjordan/islandora_workbench/assets/72275887/c290fbec-4e06-489e-8e2b-e0ba518cd490)

**Comment by mjordan** [comment]
OK, thanks, that's a bug. Do you mind sharing your config file used to create that media with the host and credentials removed? Also, did the CSV entry for that file include the ".tiff" extension or was it created from a URL that ended in "/view"?

**Comment by dara2** [comment]
In looking back to get my config file, I noticed that the place where I'm seeing this is on an older version of WB (from Dec. 2022), and also in files that I uploaded manually. When I checked files ingested using WB main branch from June, I do see .tif as the extension.

So I think that bug has already been addressed!

It sounds, then, like what I'm running into is the server timing out? I'm using `pause: 2` in my config - shouldn't that help with this type of thing? Should I try extending the pause?

**Comment by mjordan** [comment]
OK, thanks, good to know you aren't seeing any inconsistent behavior in the current Workbench.

Correct, the server is timing out on requests for certain files. `pause` is a pretty blunt instrument. I recommend `adaptive_pause` since it tracks the server's response time and pauses only when it sees the response time climb above a certain threshold. You might want to remove `pause` from your config file and try this to start with instead:

```
adaptive_pause: 2
adaptive_pause_threshold: 2.5 
```

And if that doesn't work, increase the `adaptive_pause_threshold` setting by .5 seconds (e.g. from 2.5 5o 3) until you see fewer failures.

 But, I think we need a way to make Workbench adapt better to long-timeout file downloads. That's a whole new feature. I'll open an issue for it.

**Comment by mjordan** [comment]
@dara2 can I ask, did you test any of the new config settings? I'd like to merge this branch so I can keep the PR queue moving forward. OK if you didn't, but I just wanted to check in on your testing outside of the issues you encountered above.

**Comment by dara2** [comment]
Thanks, I'll try that!
I'm in the middle of testing some of these config settings. But ran into this during check and wanted to let you know before I moved on to the actual ingest. I'll run the ingest now with the suggested pause-related settings.

**Comment by mjordan** [comment]
#471 to handle timeouts better already exists. But, I'm not sure what the best approach to take it.

**Comment by dara2** [comment]
(Moving this comment from #647 ):
Noting that this line in the config, query_csv_id_to_node_id_map_for_parents: false, resulted in child nodes not getting attached to parents.
I commented that out and am back to testing the (updated Friday) issue_613 branch.
My findings are that remote .tif files are now getting ingested, but it is taking a VERY long time to ingest each .tif. Here's what's in my config (I turned cache off altogether because I wasn't sure the state of the caching work in the branch I'm on):
```
task: create
host: "https://diglib-staging.amphilsoc.org/"
username: xxx
password: xxx
input_dir: APS
input_csv: issue_613_test.csv
output_csv: issue_613_test-output.csv
output_csv_include_input_csv: true
allow_missing_files: true
allow_adding_terms: true
validate_title_length: false
strict_check: false
standalone_media_url: true
field_for_media_title: field_pid
field_for_remote_filename: field_pid
delete_tmp_upload: true
enable_http_cache: false
adaptive_pause: 2
adaptive_pause_threshold: 2.5 
# query_csv_id_to_node_id_map_for_parents: false
media_types_override:
  - video: ['avi']
# additional_files:
 # - TEI: 139
```
Do you have any suggestions to speed up pulling over the TIF files? Creating nodes is going at a pretty good speed.

Here's a snapshot from the .log file:
```
24-Jul-23 12:06:15 - INFO - Term "Open Seadragon" (term ID 21) already exists in vocabulary "islandora_display".
24-Jul-23 12:06:15 - INFO - Term "islandora:pageCModel" (term ID 203) already exists in vocabulary "legacy_model".
24-Jul-23 12:06:16 - INFO - Term "Text" (term ID 34) already exists in vocabulary "resource_types".
24-Jul-23 12:06:16 - INFO - Term "Page" (term ID 17) already exists in vocabulary "islandora_models".
24-Jul-23 12:06:17 - INFO - Node for "Isaac Israel Hayes - #2, Page 80 (record text:302177)" created at https://diglib-staging.amphilsoc.org/node/5781.
24-Jul-23 12:07:59 - INFO - HTTP requests paused for 2 seconds because request in next log entry exceeded adaptive threshold of 2.5.
24-Jul-23 12:07:59 - INFO - {'method': 'POST', 'response': 201, 'url': '/file/upload/media/file/field_media_file?_format=json', 'response_time': 31.660573, 'average_response_time': 2.19007825}
24-Jul-23 12:08:20 - INFO - Media for https://diglib.amphilsoc.org/islandora/object/text:302177/datastream/OBJ/view created.
24-Jul-23 12:08:24 - INFO - Term "Open Seadragon" (term ID 21) already exists in vocabulary "islandora_display".
24-Jul-23 12:08:25 - INFO - Term "islandora:pageCModel" (term ID 203) already exists in vocabulary "legacy_model".
24-Jul-23 12:08:26 - INFO - Term "Text" (term ID 34) already exists in vocabulary "resource_types".
24-Jul-23 12:08:27 - INFO - Term "Page" (term ID 17) already exists in vocabulary "islandora_models".
24-Jul-23 12:08:28 - INFO - Node for "Isaac Israel Hayes - #2, Page 79 (record text:302176)" created at https://diglib-staging.amphilsoc.org/node/5782.
24-Jul-23 12:10:05 - INFO - HTTP requests paused for 2 seconds because request in next log entry exceeded adaptive threshold of 2.5.
24-Jul-23 12:10:05 - INFO - {'method': 'POST', 'response': 201, 'url': '/file/upload/media/file/field_media_file?_format=json', 'response_time': 25.68273, 'average_response_time': 3.7394469}
24-Jul-23 12:10:23 - INFO - Media for https://diglib.amphilsoc.org/islandora/object/text:302176/datastream/OBJ/view created.
24-Jul-23 12:10:26 - INFO - Term "Open Seadragon" (term ID 21) already exists in vocabulary "islandora_display".
24-Jul-23 12:10:27 - INFO - Term "islandora:pageCModel" (term ID 203) already exists in vocabulary "legacy_model".
24-Jul-23 12:10:27 - INFO - Term "Text" (term ID 34) already exists in vocabulary "resource_types".
24-Jul-23 12:10:28 - INFO - Term "Page" (term ID 17) already exists in vocabulary "islandora_models".
24-Jul-23 12:10:29 - INFO - Node for "Isaac Israel Hayes - #2, Page 78 (record text:302175)" created at https://diglib-staging.amphilsoc.org/node/5783.
24-Jul-23 12:12:38 - INFO - HTTP requests paused for 2 seconds because request in next log entry exceeded adaptive threshold of 2.5.
24-Jul-23 12:12:38 - INFO - {'method': 'POST', 'response': 201, 'url': '/file/upload/media/file/field_media_file?_format=json', 'response_time': 53.532494, 'average_response_time': 4.8909655}
24-Jul-23 12:12:56 - INFO - Media for https://diglib.amphilsoc.org/islandora/object/text:302175/datastream/OBJ/view created.
24-Jul-23 12:12:58 - INFO - Term "Open Seadragon" (term ID 21) already exists in vocabulary "islandora_display".
24-Jul-23 12:12:59 - INFO - Term "islandora:pageCModel" (term ID 203) already exists in vocabulary "legacy_model".
24-Jul-23 12:12:59 - INFO - Term "Text" (term ID 34) already exists in vocabulary "resource_types".
24-Jul-23 12:13:00 - INFO - Term "Page" (term ID 17) already exists in vocabulary "islandora_models".
24-Jul-23 12:13:01 - INFO - Node for "Isaac Israel Hayes - #2, Page 77 (record text:302174)" created at https://diglib-staging.amphilsoc.org/node/5784.
24-Jul-23 12:14:19 - INFO - Media for https://diglib.amphilsoc.org/islandora/object/text:302174/datastream/OBJ/view created.
24-Jul-23 12:14:21 - INFO - Term "Open Seadragon" (term ID 21) already exists in vocabulary "islandora_display".
24-Jul-23 12:14:22 - INFO - Term "islandora:pageCModel" (term ID 203) already exists in vocabulary "legacy_model".
24-Jul-23 12:14:22 - INFO - Term "Text" (term ID 34) already exists in vocabulary "resource_types".
24-Jul-23 12:14:23 - INFO - Term "Page" (term ID 17) already exists in vocabulary "islandora_models".
24-Jul-23 12:14:24 - INFO - Node for "Isaac Israel Hayes - #2, Page 76 (record text:302173)" created at https://diglib-staging.amphilsoc.org/node/5785.
24-Jul-23 12:15:43 - INFO - HTTP requests paused for 2 seconds because request in next log entry exceeded adaptive threshold of 2.5.
24-Jul-23 12:15:43 - INFO - {'method': 'POST', 'response': 201, 'url': '/file/upload/media/file/field_media_file?_format=json', 'response_time': 9.172282, 'average_response_time': 1.588987}
```

**Comment by dara2** [comment]
Also, I've had two of these errors so far:
```
24-Jul-23 12:17:09 - ERROR - Media not created, POST request to "/entity/media?_format=json" returned an HTTP status code of "422" and a response body of b'{"message":"Unprocessable Entity: validation failed.\\nfield_media_file.0.target_id: This value should be of the correct primitive type.\\n"}'.
```

**Comment by mjordan** [comment]
@dara2 I've added logic such that if either `query_csv_id_to_node_id_map_for_parents` or `config['csv_id_to_node_id_map_path' is false, Workbench won't use the CSV ID to node ID map at all during `--check` or during non-check. Can you pull in updates to issue_613 and retest and then we'll pick up the other threads you've noted?

**Comment by dara2** [comment]
Sure, will do that now. I am also getting this during --check on another CSV, with the issue_613 branch:
`Error: File "https://diglib.amphilsoc.org/islandora/object/text:170118/datastream/OBJ/view" in CSV row "text:170118" has an extension (html) that is not allowed in the "field_media_file" field of the "file" media type.`

I confirmed that the file (TIFF) is accessible by anonymous users and downloads when I go to the URL.

**Comment by mjordan** [comment]
Given the context (e.g URL doesn't end in .html and that we expect the file to be not HTML) I'd like to try to detect whether this is a human-friendly HTML page that the server is returning in place of a proper HTTP response code (552, etc.). I'll open an issue for this. Workbench [already does this](https://github.com/mjordan/islandora_workbench/blob/main/workbench_utils.py#L5692) when trying to access a Google Sheet, but that's possible because it's dealing with a single provider of the HTML so it's relatively easy to confirm that the page is a human-readable error message. I'm not sure we'll be able to generalize that out effectively to detect similar messages from arbitrary servers.

Is it possible for you to test against a remote Islandora server that is less likely to time out? I don't doubt Workbench is getting an HTML file for that URL, but I don't think this error is accurate (my guess is that the server is returning an HTML page saying it timed out). I could be wrong about this but because the server is returning `content-type: image/tiff` *most* of the time for that URL, it probably doesn't have an .html extension.

**Comment by dara2** [comment]
I agree that it's most likely an issue with the i7 server it's pulling from (and Workbench is reading the html error message). Because the file it errors on is different each time. I went ahead and ran the actual ingest, and interestingly it seems most of the media files are being ingested without a problem.

**Comment by mjordan** [comment]
Woo hoo!

**Comment by mjordan** [comment]
@dara2 I've pushed some work to theissue_613 branch that you'll be interested in. Details over at #647.

**Comment by mjordan** [comment]
Thanks very much @dara2. 

@dmer @ajstanley I'm eager to merge this branch so it doesn't get too out of sync. Do you want to test the other config settings as below (I've struck through the ones @dara2 has tested) before I do so or are you OK with me merging without further testing?

Media titles:

- ~`field_for_media_title`:  Identifies a CSV field name (machine name, not human readable) that will be used as the media title in `create` tasks. For example, `field_for_media_title: id`. Truncates the value of the field to 255 characters.~
- `use_nid_in_media_title`: Defaults to false. Set to true to use the parent node's node ID as the media title.
- `use_node_title_for_media_title`: Defaults to false. Set to true to use the parent node's title as the media title. Truncates the value of the field to 255 characters.

Remote filenames:

- `use_node_title_for_remote_filename`: (formerly `use_node_title_for_media_filename`) Defaults to false. Set to true to use a version of the parent node's title as the filename for a remote (http[s]) file. Replaces all non-alphanumeric characters with an underscore (`_`). Truncates the value of the field to 255 characters.
-  `use_nid_in_remote_filename`: (formerly `use_nid_in_media_filename`)  Defaults to false. Set to true to use the parent node's node ID as the filename for a remote (http[s]) file.
- ~`field_for_remote_filename`: (formerly `field_for_media_filename`) Identifies a CSV field name (machine name, not human readable) that will be used as the filename for a remote (http[s]) file. For example, `field_for_remote_filename: id`. Truncates the value of the field to 255 characters.~

Thanks!

**Comment by dmer** [comment]
@mjordan the field_for_remote_filename is the main one that I'm in need of right now and I don't have any extra time today for more testing. I vote for going ahead and merging. And thanks @dara2 for all the testing work!

**Comment by mjordan** [comment]
Merged with c3fe7e13e1ac16cd5a731555edaf376639b5513a.

**Comment by mjordan** [comment]
All new and previously undocumented config settings are now documented.
